src/abstractdatatable/join.jl

```diff
@@ -60,12 +66,20 @@ function compose_joined_table(joiner::DataTableJoiner,
     # permutation to swap rightonly and leftonly rows
     right_perm = vcat(1:ril, ril+roil+1:ril+roil+loil, ril+1:ril+roil)
     if length(leftonly_ixs) > 0
-        # compose right_perm with the permutation that restores left rows order
-        right_perm[vcat(right_ixs.join, leftonly_ixs.join)] = right_perm[1:ril+loil]
+        # combine the matched (left_ixs.orig) and non-matched (leftonly_ixs.orig) indices of the left table rows
+        # preserving the original rows order
+        all_orig_left_ixs = similar(left_ixs.orig, length(left_ixs)+length(leftonly_ixs))
+        @inbounds all_orig_left_ixs[left_ixs.join] = left_ixs.orig
+        @inbounds all_orig_left_ixs[leftonly_ixs.join] = leftonly_ixs.orig
+    else
+        # the result contains only the left rows that are matched to right rows (left_ixs)
+        all_orig_left_ixs = left_ixs.orig # no need to copy left_ixs.orig as it's not used elsewhere
     end
     all_orig_right_ixs = vcat(right_ixs.orig, rightonly_ixs.orig)
-    right_dt = DataTable(Any[resize!(col[all_orig_right_ixs], length(all_orig_right_ixs)+loil)[right_perm]
-                            for col in columns(dtr_noon)],
+    right_dt = DataTable(Any[copy!(similar(col[all_orig_right_ixs],
```

Comment:
This is actually worse than before since you call `col[all_orig_right_ixs]` repeatedly, which allocates a copy just to discard it immediately. Anyway let's leave this as it was before this commit, and improve this later with a more verbose approach.